core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointMapper.scala

```diff
@@ -0,0 +1,220 @@
+package org.typelevel.idna4s.core.uts46
+
+import scala.annotation.tailrec
+import scala.annotation.switch
+import cats.data._
+import org.typelevel.idna4s.core._
+import scala.util.control.NoStackTrace
+import java.lang.StringBuilder
+
+private[uts46] trait CodePointMapperBase {
+  protected final val VALID: Int = -1
+  protected final val VALID_NV8: Int = -2
+  protected final val VALID_XV8: Int = -3
+  protected final val IGNORED: Int = -4
+  protected final val MAPPED_MULTI_CODE_POINT = -5
+  protected final val DEVIATION = -6
+  protected final val DEVIATION_MULTI = -7
+  protected final val DISALLOWED = -8
+  protected final val DISALLOWED_STD3_VALID = -9
+  protected final val DISALLOWED_STD3_MAPPED = -10
+  protected final val DISALLOWED_STD3_MAPPED_MULTI = -11
+  protected final val DEVIATION_IGNORED = -12
+
+  protected def unsafeMapIntCodePointSentinel0(codePoint: Int): Int
+
+  protected def unsafeLookupCodePointMultiMapping0(codePoint: Int): NonEmptyList[Int]
+
+  protected def unsafeLookupDeviationCodePointMapping0(deviationCodePoint: Int): Int
+
+  protected def unsafeLookupDeviationCodePointMultiMapping0(deviationCodePoint: Int): NonEmptyList[Int]
+
+  protected def unsafeLookupDisallowedSTD3CodePointMapping0(disallowedSTD3CodePoint: Int): Int
+
+  protected def unsafeLookupDisallowedSTD3CodePointMultiMapping0(disallowedSTD3CodePoint: Int): NonEmptyList[Int]
+}
+
+object CodePointMapper extends GeneratedCodePointMapper {
+
+  /** Map the code points in the given `String` according to the UTS-46 mapping tables
+    * as described in section 5 of UTS-46. useStd3ASCIIRules is true and
+    * transitionalProcessing is false.
+    */
+  def mapCodePoints(input: String): Either[MappingFailure, String] =
```

Comment:
I wondered whether we could relax this to `CharSequence`, but it doesn't expose `codePointAt()`. I don't see why it couldn't to be honest.

Anyway, just a silly, unlikely optimization 😛

Reply:
Yeah, I've been kicking around changing it to `CharSequence`. The way the JVM's String/CharSequence APIs talk about characters, as distinct from (UTF-16) bytes, as distinct from Unicode code points, is all over the place and it makes it hard to abstract when you care both about UTF-16 bytes and Unicode code points.

What I really want is set typeclasses that allow you to inspect a StringLike values as either a sequence of bytes (in an arbitrary UTF) or distinctly as a sequence of Unicode Code points. And we could have instances for `String`, various collection types, anything which `<: CharSequence`, etc.

But, I'm trying not to get ahead of myself and that typeclass definitely shouldn't live here, though I don't know where it _should_ live.

It would be useful for idna4s, cats-uri, and case-insensitive, as well as just...in general.

Reply:
TL;DR I'm going to leave most things as `String` for now then do a pass at the end and see what can and should have a `CharSequence` or similar based API.